[X86] Invert (and X, ~(and ~Y, Z)) back into (and X, (or Y, ~Z))

Saldivarcher opened the pull request:
When `andn` is available, we should avoid switching `s &= ~(z & ~y);` into `s &= ~z | y;`

This patch turns this assembly from:
```
foo:
        not     rcx
        and     rsi, rdx
        andn    rax, rsi, rdi
        or      rcx, rdx
        and     rax, rcx
        ret
```
into:
```
foo:
        and     rsi, rdx
        andn    rcx, rdx, rcx
        andn    rax, rsi, rdi
        andn    rax, rcx, rax
        ret
```
Fixes #108731

Saldivarcher commented:
@RKSimon I wasn't sure if this was the correct approach for this, but this is the direction I went in.

Saldivarcher commented:
Here is what the assembly looks like with my changes:
```console
saldivar@pe34genoa ~/c/w/r/u/108731> cat repro.c 
#include <stdint.h>

int64_t foo(int64_t w, int64_t x, int64_t y, int64_t z) {
  int64_t s = w;
  s &= ~(x & y);
  s &= ~(z & ~y);
  return s;
}

typedef uint8_t vec16u8 __attribute__((vector_size(16)));

vec16u8 fooVec(vec16u8 w, vec16u8 x, vec16u8 y, vec16u8 z) {
  vec16u8 s = w;
  s &= ~(x & y);
  s &= ~(z & ~y);
  return s;
}
saldivar@pe34genoa ~/c/w/r/u/108731> ~/workspace/work/llvm-project.git/build_debug/bin/clang repro.c -O3 -S -masm=intel -march=znver3
saldivar@pe34genoa ~/c/w/r/u/108731> cat repro.s
        .text
        .intel_syntax noprefix
        .file   "repro.c"
        .globl  foo                             # -- Begin function foo
        .p2align        4, 0x90
        .type   foo,@function
foo:                                    # @foo
        .cfi_startproc
# %bb.0:                                # %entry
        and     rsi, rdx
        andn    rcx, rdx, rcx
        andn    rax, rsi, rdi
        andn    rax, rcx, rax
        ret
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        # -- End function
        .globl  fooVec                          # -- Begin function fooVec
        .p2align        4, 0x90
        .type   fooVec,@function
fooVec:                                 # @fooVec
        .cfi_startproc
# %bb.0:                                # %entry
        vandps  xmm1, xmm2, xmm1
        vandnps xmm2, xmm2, xmm3
        vandnps xmm0, xmm1, xmm0
        vandnps xmm0, xmm2, xmm0
        ret
.Lfunc_end1:
        .size   fooVec, .Lfunc_end1-fooVec
        .cfi_endproc
                                        # -- End function
        .ident  "clang version 20.0.0git (git@github.com:llvm/llvm-project.git f00c946c2da0caf6da4a49e87ac905a8b1d2e8b6)"
        .section        ".note.GNU-stack","",@progbits
        .addrsig
```

topperc commented:
InstCombine generally does not take subtarget aware optimizations like this. A lot of what InstCombine does is considered "canonicalization".

topperc commented:
Preventing the transform in InstCombine only solves the problem if the user was aware of the existence of andn and wrote their code as `s &= ~(z & ~y)`. A user unaware of `andn` might write `s &= ~z | y` in their code because it looks simpler.

RKSimon commented:
@Saldivarcher This needs to be done in DAGCombine, not InstCombine, to allow us to properly account for a target's support for ANDNOT instructions. I did an initial test yesterday, which hit infinite loops in VE and AArch64 targets, so to initially address #108731 we could just handle this in X86ISelLowering.

This is what I tried as an initial investigation inside DAGCombiner::visitAND:
```c
  // Fold (and X, (or Y, ~Z)) -> (and X, ~(and ~Y, Z))
  if (TLI.hasAndNot(SDValue(N, 0))) {
    SDValue X, Y, Z;
    if (sd_match(N, m_And(m_Value(X),
                          m_OneUse(m_Or(m_Value(Y), m_Not(m_Value(Z)))))))
      return DAG.getNode(
          ISD::AND, DL, VT, X,
          DAG.getNOT(
              DL, DAG.getNode(ISD::AND, DL, VT, DAG.getNOT(DL, Y, VT), Z), VT));
  
```

You could try just moving it to X86's combineAnd and replace the hasAndNot

nikic commented:
Landed https://github.com/llvm/llvm-project/commit/f1ff3a279f3320d8e0d4abbc1a8357bb51de25a2 to hopefully save the next person some time.

Saldivarcher committed:
[X86] Invert `(and X, ~(and ~Y, Z))` back into `(and X, (or Y, ~Z))`

The reason for this inversion is to utilize the `andn` instruction,
which in turn produces less assembly code.

This is the assembly we produced previously:
```
        not     rcx
        and     rsi, rdx
        andn    rax, rsi, rdi
        or      rcx, rdx
        and     rax, rcx
        ret
```

The assembly with the inversion:

```
        and     rsi, rdx
        andn    rcx, rdx, rcx
        andn    rax, rsi, rdi
        andn    rax, rcx, rax
        ret
```

Saldivarcher commented:
@topperc and @RKSimon thanks for the help and explanations, they were extremely useful. 

@nikic sorry you had to make the changes because of me! Those were good changes tho 👍

Saldivarcher commented:
@RKSimon I did pretty much what you suggested and get the results we expect:
```console
$ cat to-optimize.ll 
define dso_local i64 @foo(i64 %0, i64 %1, i64 %2, i64 %3) local_unnamed_addr {
Entry:
  %4 = and i64 %2, %1
  %5 = xor i64 %4, -1
  %6 = and i64 %5, %0
  %.not = xor i64 %3, -1
  %7 = or i64 %.not, %2
  %8 = and i64 %6, %7
  ret i64 %8
}

declare void @llvm.dbg.value(metadata, metadata, metadata) #1

define dso_local <16 x i8> @fooVec(<16 x i8> %0, <16 x i8> %1, <16 x i8> %2, <16 x i8> %3) local_unnamed_addr {
Entry:
  %4 = and <16 x i8> %2, %1
  %5 = xor <16 x i8> %4, <i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1>
  %6 = and <16 x i8> %5, %0
  %.not = xor <16 x i8> %3, <i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1, i8 -1>
  %7 = or <16 x i8> %.not, %2
  %8 = and <16 x i8> %6, %7
  ret <16 x i8> %8
}
$ ~/workspace/work/llvm-project.git/build_debug/bin/llc to-optimize.ll -o out.s --x86-asm-syntax=intel
$ cat out.s
        .text
        .intel_syntax noprefix
        .file   "to-optimize.ll"
        .globl  foo                             # -- Begin function foo
        .p2align        4, 0x90
        .type   foo,@function
foo:                                    # @foo
        .cfi_startproc
# %bb.0:                                # %Entry
        mov     rax, rcx
        and     rsi, rdx
        not     rsi
        and     rsi, rdi
        not     rax
        or      rax, rdx
        and     rax, rsi
        ret
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        # -- End function
        .globl  fooVec                          # -- Begin function fooVec
        .p2align        4, 0x90
        .type   fooVec,@function
fooVec:                                 # @fooVec
        .cfi_startproc
# %bb.0:                                # %Entry
        andps   xmm1, xmm2
        andnps  xmm1, xmm0
        andnps  xmm2, xmm3
        andnps  xmm2, xmm1
        movaps  xmm0, xmm2
        ret
.Lfunc_end1:
        .size   fooVec, .Lfunc_end1-fooVec
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
$ ~/workspace/work/llvm-project.git/build_debug/bin/llc to-optimize.ll -o out.s --x86-asm-syntax=intel -mcpu=znver3
$ cat out.s
        .text
        .intel_syntax noprefix
        .file   "to-optimize.ll"
        .globl  foo                             # -- Begin function foo
        .p2align        4, 0x90
        .type   foo,@function
foo:                                    # @foo
        .cfi_startproc
# %bb.0:                                # %Entry
        and     rsi, rdx
        andn    rcx, rdx, rcx
        andn    rax, rsi, rdi
        andn    rax, rcx, rax
        ret
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        # -- End function
        .globl  fooVec                          # -- Begin function fooVec
        .p2align        4, 0x90
        .type   fooVec,@function
fooVec:                                 # @fooVec
        .cfi_startproc
# %bb.0:                                # %Entry
        vandps  xmm1, xmm2, xmm1
        vandnps xmm2, xmm2, xmm3
        vandnps xmm0, xmm1, xmm0
        vandnps xmm0, xmm2, xmm0
        ret
.Lfunc_end1:
        .size   fooVec, .Lfunc_end1-fooVec
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```

Saldivarcher commented:
@RKSimon ping!

RKSimon commented:
@Saldivarcher sorry this fell off my review list as I wasn't a listed reviewer - will take another look soon

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -50034,6 +50034,32 @@ static bool hasBZHI(const X86Subtarget &Subtarget, MVT VT) {
+///    `(and X, (or Y, ~Z))`
+///
+/// But we should undo this transformation if the `andn` instruction is
+/// available to us.
```
Better to rephrase this in explicit terms of what this fold does:
```
Folds (and X, (or Y, ~Z)) --> (and X, ~(and ~Y, Z))
This undoes the inverse fold performed in InstCombine
```

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -50034,6 +50034,32 @@ static bool hasBZHI(const X86Subtarget &Subtarget, MVT VT) {
+/// But we should undo this transformation if the `andn` instruction is
+/// available to us.
+static SDValue combineAndNotOrIntoAndNotAnd(SDNode *N, SelectionDAG &DAG,
+                                            const X86Subtarget &Subtarget) {
```
Is Subtarget needed?

goldsteinn replied:
Think there is a missing check for whether andn is supported...

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -50034,6 +50034,32 @@ static bool hasBZHI(const X86Subtarget &Subtarget, MVT VT) {
+  MVT VT = N->getSimpleValueType(0);
+  SDLoc DL(N);
+  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
+  if (TLI.hasAndNot(SDValue(N, 0))) {
```
(style)
```
if (!TLI.hasAndNot(SDValue(N, 0)))
  return SDValue();
```

Saldivarcher replied:
So the latter is preferred?
```c++
static SDValue combineAndNotOrIntoAndNotAnd(SDNode *N, SelectionDAG &DAG) {

  using namespace llvm::SDPatternMatch;
  MVT VT = N->getSimpleValueType(0);
  SDLoc DL(N);
  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
  if (TLI.hasAndNot(SDValue(N, 0))) {
    SDValue X, Y, Z;
    if (sd_match(N, m_And(m_Value(X),
                          m_OneUse(m_Or(m_Value(Y), m_Not(m_Value(Z)))))))
      return DAG.getNode(
          ISD::AND, DL, VT, X,
          DAG.getNOT(
              DL, DAG.getNode(ISD::AND, DL, VT, DAG.getNOT(DL, Y, VT), Z), VT));
  }
  return SDValue();
}
```
vs
```cpp
static SDValue combineAndNotOrIntoAndNotAnd(SDNode *N, SelectionDAG &DAG) {

  using namespace llvm::SDPatternMatch;
  MVT VT = N->getSimpleValueType(0);
  SDLoc DL(N);
  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
  if (!TLI.hasAndNot(SDValue(N, 0)))
    return SDValue();

  SDValue X, Y, Z;
  if (sd_match(
          N, m_And(m_Value(X), m_OneUse(m_Or(m_Value(Y), m_Not(m_Value(Z)))))))
    return DAG.getNode(
        ISD::AND, DL, VT, X,
        DAG.getNOT(DL, DAG.getNode(ISD::AND, DL, VT, DAG.getNOT(DL, Y, VT), Z),
                   VT));

  return SDValue();
}
```

RKSimon replied:
Yes - the idea is to early-out from functions instead of nesting too many if()s
https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Saldivarcher replied:
That makes sense, thanks for sharing that!

Saldivarcher replied:
The latter code is already in the PR btw 👍

RKSimon commented on llvm/test/CodeGen/X86/avx512vl-logic.ll:
```diff
@@ -980,7 +980,7 @@ define <4 x i32> @ternlog_or_andn(<4 x i32> %x, <4 x i32> %y, <4 x i32> %z) {
 ; CHECK-LABEL: ternlog_and_orn:
 ; CHECK:       ## %bb.0:
-; CHECK-NEXT:    vpternlogd $176, %xmm1, %xmm2, %xmm0
+; CHECK-NEXT:    vpternlogd $208, %xmm2, %xmm1, %xmm0
```
rebase + regenerate this file

Saldivarcher replied:
Looks like someone made this more generic.

RKSimon commented on llvm/test/CodeGen/X86/pr108731.ll:
```diff
@@ -0,0 +1,40 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-gnu-unknown -mcpu=znver3 | FileCheck %s
```
Make test more generic - replace znver3 with x86-64-v3 and add a x86-64-v2 test for case without BMI1 support:
```
RUN: llc < %s -mtriple=x86_64-- -mcpu=x84-64-v2 | FileCheck %s --check-prefixes=CHECK,NOBMI
RUN: llc < %s -mtriple=x86_64-- -mcpu=x84-64-v3 | FileCheck %s --check-prefixes=CHECK,BMI
```

Saldivarcher replied:
Didn't know about `x86-64-v3` or `x86-64-v2`, thank you!

RKSimon commented on llvm/test/CodeGen/X86/pr108731.ll:
```diff
@@ -0,0 +1,40 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-gnu-unknown -mcpu=znver3 | FileCheck %s
+
+define dso_local i64 @foo(i64 %0, i64 %1, i64 %2, i64 %3) local_unnamed_addr {
```
Test cleanup (same for fooVec):
remove dso_local  / local_unnamed_addr noise
use named variables not numbers
rename foo / fooVec with descriptive names

Saldivarcher replied:
Hopefully the names I chose are ok 👍

RKSimon commented on llvm/test/CodeGen/X86/pr108731.ll:
```diff
@@ -0,0 +1,40 @@
+  ret i64 %8
+}
+
+declare void @llvm.dbg.value(metadata, metadata, metadata) #1
```
remove unnecessary declaration

RKSimon reviewed (approved):
LGTM - cheers

Saldivarcher commented:
@RKSimon can you rebase and merge the commit for me, I don't have write access. Thank you!

bgra8 commented:
@Saldivarcher this patch is causing a huge amount (hundreds) of breakages in our codebase at google. (the compiler enters an infinite loop and never finishes)

Here's a reproducer:

```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-generic-linux-gnu"

define void @_renamed(i64 %error_code.coerce) #0 {
entry:
  %0 = and i64 %error_code.coerce, 4294967296
  %1 = xor i64 %0, -1
  %2 = and i64 %error_code.coerce, %1
  %3 = icmp eq i64 %2, 0
  br i1 %3, label %5, label %4

4:                                                ; preds = %entry
  unreachable

5:                                                ; preds = %entry
  ret void
}

attributes #0 = { "target-cpu"="haswell" }
```

This compiles in under a second with the previous version and never seems to finish with this patch applied.

Can you please revert while investigating?

Saldivarcher commented:
@bgra8 actually it looks like this commit fixes the issue:
```
commit ec78f0da0e9b1b8e2b2323e434ea742e272dd913
Author: Simon Pilgrim <llvm-dev@redking.me.uk>
Date:   Tue Oct 15 16:54:08 2024 +0100

    [X86] combineAndNotOrIntoAndNotAnd - don't attempt with constant operands
    
    Don't fold AND(X,OR(NOT(Z),C)) -> AND(X,NOT(AND(Z,C'))) as DAGCombiner will invert it back again.
    
    Fixes #112347 
```

Can you give it a try?

dyung commented:
@Saldivarcher we are also seeing compiler timeouts on an internal test which I bisected back to your commit which ec78f0da0e9b1b8e2b2323e434ea742e272dd913 does not fix. I have filed the details in #113240, can you take a look?

Saldivarcher commented:
#113264 also fixes some infinite loops.

bgra8 commented:
Commit ec78f0da0e9b1b8e2b2323e434ea742e272dd913 does not help for most of the failing targets on our side.

But https://github.com/llvm/llvm-project/pull/113264 fixes all targets we've identified so far.
